### PR TITLE
Add Cloud Object Translators for Coders

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
These are all the translators with known types.

Implementations and tests for standard coders that aren't known types in Dataflow are still in progress.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  static final String KIND_GLOBAL_WINDOW = "kind:global_window";
+  static final String KIND_INTERVAL_WINDOW = "kind:interval_window";
+  static final String KIND_BYTES = "kind:bytes";
+  static final String KIND_VARINT = "kind:varint";
+  static final String KIND_LENGTH_PREFIX = "kind:length_prefix";
+  static final String KIND_PAIR = "kind:pair";
+  static final String KIND_STREAM = "kind:stream";
+  static final String KIND_WINDOWED_VALUE = "kind:windowed_value";
+  static final String KIND_WINDOWED_VALUE_ONLY_WRAPPER =
+      CloudObject.forClass(WindowedValue.ValueOnlyWindowedValueCoder.class).getClassName();
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java]: We should drop ValueOnlyWindowedValueCoder as a special CloudObjectKind and turn ValueOnlyWindowedValueCoder into a CustomCoder.

Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java]: Pending https://github.com/apache/beam/pull/2685

Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectKinds.java]: #2685 is merged.

**Comment by tgroh** [comment]
Rebased, PTAL

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java [hunk 2]
```diff
@@ -34,4 +34,10 @@
    * Converts back into the original object from a provided {@link CloudObject}.
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java]: An translator that takes -> A translator that takes

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java [hunk 3]
```diff
@@ -34,4 +34,10 @@
...
    */
   T fromCloudObject(CloudObject cloudObject);
+
+  /**
+   * Gets the class name that will represent any {@link CloudObject} created by this {@link
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java]: represent any -> represent the

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java [hunk 4]
```diff
@@ -30,16 +64,80 @@
...
+
+  // TODO: Coders with a custom wire format.
+  // SerializableCoder.class
+  // AvroCoder.class
+  // ProtoCoder.class
   @Override
-  public Map<String, CloudObjectTranslator<? extends Coder>> classNamesToTranslators() {
-    // TODO: Add translators
-    return Collections.emptyMap();
+  publicO
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: public0?

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+  /**
+   * Returns a {@link CloudObjectTranslator} that produces a {@link CloudObject} that is of kind
+   * "bytes".
+   */
+  static CloudObjectTranslator<ByteArrayCoder> bytes() {
+    return new CloudObjectTranslator<ByteArrayCoder>() {
+      @Override
+      public CloudObject toCloudObject(ByteArrayCoder target) {
+        return addComponents(
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java]: can we make this kind:bytes

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  /**
+   * Returns a {@link CloudObjectTranslator} that produces a {@link CloudObject} that is of kind
+   * "varint".
+   */
+  static CloudObjectTranslator<VarLongCoder> varInt() {
+    return new CloudObjectTranslator<VarLongCoder>() {
+      @Override
+      public CloudObject toCloudObject(VarLongCoder target) {
+        return addComponents(
+            CloudObject.forClass(target.getClass()), Collections.<Coder<?>>emptyList());
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java]: kind:varint

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java [hunk 7]
```diff
@@ -50,7 +50,7 @@
  *       expensive.</li>
  *   <li>{@link #getEncodingId} and {@link #getAllowedEncodings}: by default, the encoding id
  *       is the empty string, so only the canonical name of the subclass will be used for
- *       compatibility checks, and no other encoding ids are allowed.</li>
+ *       compatibility checks, and no other encoding ids al
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java]: ???

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java [hunk 8]
```diff
@@ -30,16 +64,80 @@
...
+              Maps.asMap(
+                  KNOWN_ATOMIC_CODERS,
+                  new Function<Class<? extends Coder>, CloudObjectTranslator<? extends Coder>>() {
+                    @Override
+                    public CloudObjectTranslator<? extends Coder> apply(
+                        Class<? extends Coder> input) {
+                      return CloudObjectTranslators.atomic(input);
+                    }
+                  }));
+  // TODO: Atomic, GCPIO Coders:
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: Fix spacing on comment if you want to keep it.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java [hunk 9]
```diff
@@ -34,4 +34,10 @@
...
+   * CloudObjectTranslator}.
+   */
+  String cloudObjectClassName();
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java]: Typically you would have the registrar provide the mapping of string -> translator and class -> translator and not have the translator self report half of the mapping.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java (hunk 8 above)
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: Rm'd.

Sounds vaguely threatening.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java (hunk 7 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StandardCoder.java]: Whoops.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java (hunk 9 above)
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslator.java]: Moved into the translator.

**Pull request closed without merging**